### PR TITLE
[BEAM-986] Improve ReduceFnRunner prefetching

### PR DESCRIPTION
**Pull request by scwhittle**
Hi @kennknowles , can you please take a look?

This refactoring ensures that for batches of elements or timers we have at most 2 fetches, one to prefetch the should fire bits and one to fetch the actual firing state.  The current behavior can end up with 2 fetches per window triggered.

### CODE REVIEW

**Comment by scwhittle** [comment]
The test failure looks unrelated to me and it passes locally for me. Is that a known flaky test or should I take a closer look?

**Comment by kennknowles** [comment]
Yea, not worried about the Travis failure. But Jenkins did not run, probably due to some transient infrastructure issue. Can you rebase and force push to the branch? That will kick Jenkins again.

**Comment by scwhittle** [comment]
Fixed some checkstyle issues

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 1]
```diff
@@ -565,98 +591,146 @@ public void onMerge(Collection<W> toBeMerged, W mergeResult) throws Exception {
...
+    public boolean windowIsActiveAndOpen() {
+      return activeWindows.isActive(directContext.window())
+          && !triggerRunner.isClosed(directContext.state());
+    }
+  }
 
-    if (!windowIsActiveAndOpen) {
-      WindowTracing.debug(
-          "ReduceFnRunner.onTimer: Note that timer {} is for non-ACTIVE window {}", timer, window);
+  public void onTimers(Iterable<TimerData> timers) throws Exception {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: I've reviewed up to here.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 2]
```diff
@@ -303,14 +309,34 @@ public void processElements(Iterable<WindowedValue<InputT>> values) throws Excep
...
 
-    // Trigger output from any window for which the trigger is ready
-    for (W mergedWindow : windowsToConsider) {
-      ReduceFn<K, InputT, OutputT, W>.Context directContext =
-          contextFactory.base(mergedWindow, StateStyle.DIRECT);
-      ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-          contextFactory.base(mergedWindow, StateStyle.RENAMED);
-      triggerRunner.prefetchShouldFire(mergedWindow, directContext.state());
-      emitIfAppropriate(directContext, renamedContext);
+    if (!windowsToConsider.isEmpty()) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: For readability, it would be nice for this to be an early return, too.

But actually all the loops are dead if the condition holds so I'd be inclined leave out the overarching `if` unless you know allocating an iterator is expensive, which it isn't here.

Since these are smallish in-memory loops, it isn't expensive to re-iterate and it could help readability a lot. Consider if you broke each loop out into methods:

 - `void prefetchShouldFire(Collection<W>)`
 - `Collection<W> windowsThatShouldFire(Collection<W>)`
 - `void prefetchEmitPane(Collection<W>)` (here you've merged with above)
 - `void emitPanes(Collection<W>)`.

I'm trying to keep this method under control, as the class is quite difficult already.

Comment (reuvenlax) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Just a drive-by comment, but are we sure reiteration is cheap here? This appears to be a per-element cost, and given the scale we've found that introducing even small extra per-element cost can be quite noticeable.

Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: It is a per-bundle cost of iterating in-memory O(<# of windows present in bundle, post-merge>) so it scales well with bundle size. Even if it scaled less well, I'd expect this to be unmeasurable relative to even to other overheads, much less nontrivial processing. FWIW my proposal changes the number of iterations from 3 to 4; I'm also OK with leaving it at 3 and merging `prefetchEmitPane` into `windowsThatShouldFire` which should elide any disagreement.

If we have measurements demonstrating this cost, then certainly we should document them and leave micro-optimizations in place. But given where the SDK is today, let's favor readability and abstraction over micro optimizations unless we have compelling evidence.

Comment (reuvenlax) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Sorry, I didn't properly explain myself:

In the case of small windows and high skew from a source, the number of windows in a bundle is ~ number of elements in the bundle. This is not a great situation for many other reasons, but we empirically see this happening a lot for various sources.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/PaneInfoTracker.java [hunk 3]
```diff
@@ -55,6 +55,12 @@ public void clear(StateAccessor<?> state) {
     state.access(PANE_INFO_TAG).clear();
   }
 
+  @SuppressFBWarnings(value = "RV_RETURN_VALUE_IGNORED_NO_SIDE_EFFECT",
```
Comment (scwhittle) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/PaneInfoTracker.java]: Checkstyle and builds work without this so I removed it.  I had copied it from somewhere else but perhaps it's no longer necessary.

**Comment by bjchambers** [comment]
I think one of the reasons we didn't do this was concern about the potential memory impact -- doing a window at a time means we don't need all the state for all the windows to fit in memory at once. Have we measured that (1) this necessary and (2) this isn't going to cause additional memory pressure?

**Comment by scwhittle** [comment]
I have observed up to a 15% improvement in a nexmark benchmark where we
receive bundles of 10 timers at a time etc and without this we end up doing
10 serial get data requests for the key.
I think that memory impact is bounded since only prefetch up to 1MB from
each tag list and paginate the rest upon demand and thus doesn't seem like
to large a concern (or at least not any more than many timers firing for
many different keys at the same time).

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java (hunk 2 above)
Comment (scwhittle) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: I've refactored things some to make the flow more straightforward.
I moved some of the window stuff from processElement to improve their prefetching and also avoid needing to build up the consideration set upon what processElement returns.
I left some of the loops in the function since they were just a single line and a function for them seemed verbose, let me know if you'd still like them separated.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 4]
```diff
@@ -325,14 +369,23 @@ public void persist() {
...
+    Map<W, W> windowToMergeResult = new HashMap<>();
+
     // No-op if no merging can take place
     if (windowingStrategy.getWindowFn().isNonMerging()) {
-      return ImmutableMap.of();
+      for (WindowedValue<?> value : values) {
+        for (BoundedWindow untypedWindow : value.getWindows()) {
+          @SuppressWarnings("unchecked")
+          W window = (W) untypedWindow;
+          windowToMergeResult.put(window, window);
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: If iteration and allocation of O(<# windows>) is a concern, then you could implement a trivial `IdentityMap` here. The only windows that will be requested will be from the input, so you don't need to even check membership before returning the input to `get(W)`.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 5]
```diff
@@ -471,38 +526,46 @@ public void onMerge(Collection<W> toBeMerged, W mergeResult) throws Exception {
...
+  private ImmutableSet<W> toMergedWindows(final Map<W, W> windowToMergeResult,
+      final Collection<? extends BoundedWindow> windows) {
+    return ImmutableSet.copyOf(
+        FluentIterable.from(windows).transform(
+            new Function<BoundedWindow, W>() {
+              @Override
+              public W apply(BoundedWindow untypedWindow) {
+                @SuppressWarnings("unchecked")
+                W window = (W) untypedWindow;
+                return checkNotNull(windowToMergeResult.get(window));
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Add an error message that will direct a user if this happens. Something like "internal error in the Beam SDK" and/or "inconsistent state; this is definitely a bug in the SDK that should be reported".

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 6]
```diff
@@ -561,102 +623,148 @@ public void onMerge(Collection<W> toBeMerged, W mergeResult) throws Exception {
...
-    checkArgument(timer.getNamespace() instanceof WindowNamespace,
-        "Expected timer to be in WindowNamespace, but was in %s", timer.getNamespace());
-    @SuppressWarnings("unchecked")
-    WindowNamespace<W> windowNamespace = (WindowNamespace<W>) timer.getNamespace();
-    W window = windowNamespace.getWindow();
-    ReduceFn<K, InputT, OutputT, W>.Context directContext =
-        contextFactory.base(window, StateStyle.DIRECT);
-    ReduceFn<K, InputT, OutputT, W>.Context renamedContext =
-        contextFactory.base(window, StateStyle.RENAMED);
+  private class TimerContext {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Can you document this a bit more? We have a lot of `XYZContext` classes floating around, and they don't all have an obvious meaning.

This actually looks it could be an `EnrichedTimer` as it basically adds some queries that you can ask about a particular `TimerData`.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 7]
```diff
@@ -561,102 +623,148 @@ public void onMerge(Collection<W> toBeMerged, W mergeResult) throws Exception {
...
-          key, window, timer.getTimestamp(), timerInternals.currentInputWatermarkTime(),
-          timerInternals.currentOutputWatermarkTime());
-      if (windowIsActiveAndOpen) {
-        emitIfAppropriate(directContext, renamedContext);
+    // For those windows that are active and open, prefetch the triggering or emitting state.
+    for (TimerContext timerContext : timerContexts) {
+      if (timerContext.windowIsActiveAndOpen()) {
+        ReduceFn<K, InputT, OutputT, W>.Context directContext = timerContext.directContext;
+        if (timerContext.isGarbageCollection) {
+          prefetchOnTrigger(directContext, timerContext.renamedContext);
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Is this legitimate? At GC time, we should definitely be prepared to emit a pane if the closing behavior says so. When I look at what `prefetchOnTrigger` does, I think only `watermarkHold.prefetchExtract(...)` is actually unconditionally needed.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java [hunk 8]
```diff
@@ -783,7 +897,7 @@ private void emitIfAppropriate(ReduceFn<K, InputT, OutputT, W>.Context directCon
   /**
    * Do we need to emit a pane?
    */
-  private boolean needToEmit(boolean isEmpty, boolean isFinished, PaneInfo.Timing timing) {
+  private boolean needToEmitPane(boolean isEmpty, boolean isFinished, PaneInfo.Timing timing) {
```
Comment (kennknowles) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: To be honest, the term "pane" misleads users pretty often - they don't realize that they are "just elements" so expect things like `Combine.PerKey()` to have a unique value per key even in the presence of multiple "panes". So I'd rather not propagate it, but rather start to eliminate it where we can.

**Comment by kennknowles** [comment]
Noting that the build is actually "green" and the failure is unrelated issue in the apex runner dependencies. If you rebase atop recent changes it should clear up.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java (hunk 4 above)
Comment (scwhittle) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Previously I was using the value set of the map to determine the windows present so this wouldn't have worked.  However I removed this by separating collectAndMergeWindows into two functions collectWindows and mergeWindows.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java (hunk 6 above)
Comment (scwhittle) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: Went with EnrichedTimerData

File: runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java (hunk 7 above)
Comment (scwhittle) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java]: This looks correct to me, prefetchOnTrigger prefetches all the state that onTrigger uses.  It's possible that function could be additionally refactored to reduce state read if isFinished is true but I'd rather not do that in this cl since I think it would need to reach into the PaneInfoTracker etc.

The garbage collection path calls onTrigger so we don't need to call prefetchEmit which fetches some additional state.

**Comment by scwhittle** [comment]
I'm running a performance test to make sure things still look good with the updated PR

**Comment by kennknowles** [comment]
LGTM and merging.
